### PR TITLE
Fix shapes interactivity with scale != 1 (selection, rotate/resize)

### LINKED ISSUE #4268: Shapes layer with high `scale` values is hard to interact with in the GUI
## 🐛 Bug

A shapes layer with high `scale` values is hard to modify in the GUI because at some point the blue control points of the selected shape cannot be clearly seen anymore. I tried to click on some of the control points, but I could not manager to move any of them.

Similarly drawing a shape on a shapes layer with a high `scale` value is very difficult, likely for similar reasons.

This seems to be a combination of `scale` values and the current zoom of the camera. I.e. higher scale values and being more zoomed out both contribute to this problem.

Here's a screenshot of what this can look like after hovering (and probably selecting) a shape.

Screen Shot 2022-03-15 at 2 24 09 PM (2)

## To Reproduce

Here's a code example that causes the issue on my setup.

```python
import napari

shapes_data = [[
    [0, 0],
    [0, 10],
    [10, 10],
    [10, 0],
]]
scale = (1000, 1000)
viewer = napari.Viewer()
shapes = viewer.add_shapes(
    data=shapes_data,
    scale=scale,
)
napari.run()
```

## Expected behavior

The appearance of the blue selection box and control points are independent of the scale of the layer, and maybe the zoom of the camera.

## Environment

```
napari: 0.4.14.dev135+gc8a62b43.d20220314
Platform: macOS-10.16-x86_64-i386-64bit
System: MacOS 11.6.2
Python: 3.9.6 (default, Jul 30 2021, 09:31:09) [Clang 10.0.0 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.0
SciPy: 1.7.1
Dask: 2021.07.2
VisPy: 0.9.6

OpenGL:
- GL version: 2.1 INTEL-16.5.7
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1440x900, scale 2.0
- screen 2: resolution 2560x1080, scale 1.0

Plugins:
- console: 0.0.4
- scikit-image: 0.4.14.dev135+gc8a62b43.d20220314
```

### LINKED ISSUE #4538: Shapes Layer: Unable to draw horizontal/verticle line
## 🐛 Bug

It seems impossible to draw a horizontal or vertical line.
Perhaps related to:
https://github.com/napari/napari/pull/3259

## To Reproduce

1. Open napari
2. Add shapes layer
3. Try to draw a horizontal or tercile line

https://user-images.githubusercontent.com/76622105/168588334-68040e6d-2943-427e-9f92-7c66f644a627.mp4

You can see the line skipping from almost horizontal to the "other side" of horizontal. Same with vertical. I can never just make a flat or vertical line.

I thought maybe using shift would help, but that's actually...worse!
Holding shift when starting to draw yields lines fixed at the SE direction. 
(Pressing shift while drawing locks the current orientation).

https://user-images.githubusercontent.com/76622105/168588558-fee849f8-ce50-4221-bb12-e2daa8771df1.mp4

## Expected behavior

It should be possible to draw lines oriented at the 0°/180° or 90°/270° orientations free-hand.
Also, shift before drawing should "skip" through 45° steps all the way from 0° to 360°.

## Environment

```
napari: 0.4.16rc2.dev17+g0c148bb6
Platform: macOS-12.3.1-arm64-arm-64bit
System: MacOS 12.3.1
Python: 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:14) [Clang 12.0.1 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.21.5
SciPy: 1.8.0
Dask: 2022.02.0
VisPy: 0.10.0

OpenGL:
- GL version: 2.1 Metal - 76.3
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1680x1050, scale 2.0

Plugins:
- aicsimageio-in-memory: 0.6.1
- aicsimageio-out-of-memory: 0.6.1
- clEsperanto: 0.16.0
- console: 0.0.4
- napari-accelerated-pixel-and-object-classification: 0.6.8
- napari-assistant: 0.1.7
- napari-math: 0.0.1b0
- napari-ome-zarr: 0.4.0
- napari-sklearn-decomposition: 0.0.1
- napari-svg: 0.1.6
- napari-time-slicer: 0.4.3
- napari-tools-menu: 0.1.10
- napari_skimage_regionprops1: 0.4.0
- napari_skimage_regionprops2: 0.4.0
- natari: 0.2.7
- ome-types: 0.2.9
- scikit-image: 0.4.16rc2.dev17+g0c148bb6
- stardist-napari: 2022.4.8
```

### LINKED ISSUE #5752: Selection outline should not be affected by `scale`
## 🐛 Bug

While creating new shapes, the shape is marked as selected. Shis adds an outline which seems to be affected by the layer scale and zoom in an odd way. With big scales, this makes picking on some image feature really hard, because the line is covering the features.

## To Reproduce

```py
import numpy as np
import napari
v = napari.Viewer()
sl = v.add_shapes(scale=[100, 100])
sl.add_paths([[0, 0], [1, 1], [3, 4]])
sl.selected_data = {0}
sl.mode = 'select'
```

Then manually zoom out.

## Expected behavior

The outline remains the same size in data space? Or maybe even in screen space?

### PR DESCRIPTION
**Pull request by brisvag**
# Description

Scale layer interactivity and visualisation are currently not accounting for the layer scale.

Try the following:

```py
import numpy as np
import napari
v = napari.Viewer()
il = v.add_image(np.random.rand(100, 100))
sl = v.add_shapes([[0, 0], [1, 1]], shape_type='path', scale=[100, 100], edge_width=0.1)
```

You'll see a few issues:
- highlight line is not in screen space as advertised. (This is actually a problem with points as well)
- the `rotate` handle is way too far because its position is also calculated not accounting for scale
- interacting with the shape in any mode is basically impossible, cause coordinates do not account for the scaling of the layer.

This PR basically adds a `/ layer.scale[-1]` in several places to get back into correct "sceen space". Note that I used the last dimension as we cannot do anisotropic sizes, just like we recently chose to do with points in #5582.

This PR also fixes #4538 by changing the logic of how the "minimum drag" is calculated.

PS: ideally we want to transition to using the `SelectionOverlay` in the future, but this is a much bigger effort.
 
Also fixes #5752.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
@brisvag very odd, but zoom out, grab the rotation handle, and try move mouse away while trying to rotate and you will find the point at which it will rotate. So it's the graphic of the handle not matching the actual grab/interaction point.

**Comment by brisvag** [comment]
> @brisvag very odd, but zoom out, grab the rotation handle, and try move mouse away while trying to rotate and you will find the point at which it will rotate. So it's the graphic of the handle not matching the actual grab/interaction point.

The intection point matches correctly, cause it gets detected by `get_value`, which works fine. It's just the math afterwards about the angles that's not working correctly. Or did you manage to rotate by starting the click elsewhere instead of the rotation handle? o.o

**Comment by brisvag** [comment]
I have a feeling this is the same as #4538, but exacerbated by the scale. This would result in the minumum "allowed" box dimensions being bigger. So you have to move further away from the shape in order for anything to happen.

**Comment by brisvag** [comment]
Ok, I think I got it; it's similar to that problem, but not exactly the same. I also fixed that one though, so that's nice :P (fixes #4538)

**Comment by psobolewskiPhD** [comment]
Gah! Nice!
Edit2: I figured it out. DERP.
Still don't get why `elif vertex == 8` is the drag handle, when `Box.HANDLE` is 9.

Edit: also use of `scale` under `elif vertex < Box.LEN` is pretty 👀 when it's not related to layer scale 🤣 Maybe should just rename that to reduce confusion with the scale of the layer?

**Comment by brisvag** [comment]
> While the rotation and handle are under `elif vertex == 8` (I'd have expected Box.HANDLE here, but that's 9 exploding_head), so why does that code trigger and play a role?

OEF.... You're right, this is all off :/ One more reason to replace it with the overlay. I wouldn't touch it here though, since it's working ^^'
 
> Edit: also use of `scale` there is pretty :eyes: when it's not related to layer scale :rofl:

You mean specifically in the `_move` callback? Sure, that's the scaling of the resizing... I renamed it `drag_scale`, maybe clearer?

**Comment by brisvag** [comment]
> Still don't get why `elif vertex == 8` is the drag handle, when `Box.HANDLE` is 9.

I guess the indexing starts from 0 and there are 8 "normal" handles, so `8` is actually the handle. The `Box` constant is probably wrong, and the callback is double-wrong to compensate :P

**Comment by brisvag** [comment]
I doubt the mac test is related; @psobolewskiPhD have you seen this elsewhere?

**Comment by psobolewskiPhD** [comment]
I'm not sure what's up with the plugin test. I've seen it before and a re-run usually fixes...

I think edge width still behaves oddly with this fix? Not intuitively for me anyways. If you don't set it like you did in the example code (so it uses default) it's impossibly thick. If you use your example, then the slider is useless. 
I think that line edge thickness shouldn't involve scale, just screen pixels so that they are visible?

**Comment by brisvag** [comment]
That's not a new behaviour; the same happens in main. I also think it's "correct": edge width is treated the same as in vectors, and the same as `size` is treated in points. They are values in data coordinates. This means that a line of thickness 1 which goes from `(0, 0)` to `(0, 1)` is going to be basically a square.

this could be changed arguably, but I'd say it's out of scope for this PR and would require some design discussion.

**Comment by brisvag** [comment]
@psobolewskiPhD friendly reminder :) Do you find my above explanation convincing?

File: napari/layers/shapes/shapes.py [hunk 1]
```diff
@@ -2617,12 +2628,24 @@ def _transform_box(self, transform, center=(0, 0)):
...
+                self._rotation_handle_length
+                * self.scale_factor
+                / self.scale[-1]
+            )
             handle_vec = box[Box.HANDLE] - box[Box.TOP_CENTER]
             cur_len = np.linalg.norm(handle_vec)
             box[Box.HANDLE] = box[Box.TOP_CENTER] + r * handle_vec / cur_len
         self._selected_box = box + center
 
+    def _update_draw(
```
Comment (psobolewskiPhD) [inline napari/layers/shapes/shapes.py:2680]: @brisvag I think I get everything, but I don't get this—it doesn't seem used?
Can you clarify for my edification?

**Review by psobolewskiPhD — APPROVED** [review]
It's odd to see so many `layer.scale_factor / layer.scale[-1]` and `self.scale_factor / self.scale[-1]` but I don't see a clear alternative?

Can you update the OP to clarify that this also fixes https://github.com/napari/napari/issues/4538
and delete bits about the rotation not working, since it does work now 😃

**Comment by Czaki** [comment]
could you fix tests before review?

File: napari/layers/shapes/shapes.py (hunk 1 above)
Comment (brisvag) [reply napari/layers/shapes/shapes.py:2680]: This is overloading the base layer. It's called every time the zoom level changes, which updates `self.scale_factor`. We need this change to propagate to the highlight (depending on zoom we need to decrease the size of the mesh that makes up the highlight)

Comment (psobolewskiPhD) [reply napari/layers/shapes/shapes.py:2680]: Oh! I see! Thanks! 😊

**Comment by brisvag** [comment]
Yeah, those might be real fails. I'm looking into it.

**Comment by brisvag** [comment]
Ok, found the issue with the test: since now `get_value` depends on the zoom level, we need to initialize the `scaling_factor` to a sensible value (otherwise it's equivalent to supidly zoomed out) so that the vertice do not in practice overlap each other.

This is a bit of a downside to this approach, because now `get_value` is dependent on the state of the layer (because the "size" of the vertices is determined by zoom). @jni I'm sure you have opinions about this, but I think the current behaviour (i.e: usability in the gui) is worth the cost.

An alternative would be to have vertices have a *fixed* size in world space, but that's a whole new can of worms.

---

PS: Most of the changes to tests are actually just to simplify the event creation to mirror my changes to point tests in #5742. The only relevant changes are:
- setting the scale factor in `create_known_shapes_layer`
- fixing the currently broken test `test_drag_vertex` which was dragging the wrong vertex (due to the aformentioned overlapping vertices).

**Comment by jni** [comment]
@brisvag sorry, could you summarise the whole thing (maybe in the original description comment since that'll be the commit message)? I don't really understand what you said — get_status generally should be related to the screen->data transform and therefore affected by the zoom level?

Ultimately skimming the code I'm unhappy about all of the manual transforms (`/ scale`) as opposed to the sort of thing proposed in #3783, but I think that's vaguely unavoidable in the current code base. I also don't know what you mean by not being able to handle anisotropic shapes — that seems like a huge problem? 😅

**Comment by brisvag** [comment]
> get_status generally should be related to the screen->data transform and therefore affected by the zoom level?

Well, to me `Layer.get_value` presents itself as a "pure" method that just inspects the data, so maybe it's unintuitive that it depends on the state of the viewer the layer is in... But yeah, not blocking unless someone thinks it is :P

> Ultimately skimming the code I'm unhappy about all of the manual transforms (/ scale) as opposed to the sort of thing proposed in https://github.com/napari/napari/issues/3783, but I think that's vaguely unavoidable in the current code base.

Yeah, I don't think it's worth blocking this fix for that, but it would definitely be nicer to have exoplicit transforms for these things.

> I also don't know what you mean by not being able to handle anisotropic shapes — that seems like a huge problem? sweat_smile

Not anisotropic *shapes*, but `Shapes` with anisotropic *scale*. If you change the example in the top post to have scale `(100, 50)`, you get strange stuff (but the same is true for main). The main issue here is that this also results in the "interaction box vertices" being scaled in size, which makes interacting with them odd (their real size is different from the apparent size) this is also true in main.


PS: This is for some reason still not working as I intended with scales... I need to check a couple things.

**Comment by brisvag** [comment]
Soemthing's definitely still off with mouse scaled get_value... made into draft for now to avoid merging until I figure it out.

**Comment by psobolewskiPhD** [comment]
@brisvag 
Do you think you could split off the fix to https://github.com/napari/napari/issues/4538
I think it should be pretty easy to review and merge.

**Comment by brisvag** [comment]
Done!

This one still needs fixing, so hopefully that part can get in.

File: napari/layers/shapes/_tests/test_shapes_mouse_bindings.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import collections
+from dataclasses import dataclass, field
+from typing import List, Optional, Tuple, Union
```
Comment (jni) [inline napari/layers/shapes/_tests/test_shapes_mouse_bindings.py]: @brisvag sorry to be a pain, but I feel like I want to pull out these changes, which as far as I can tell have nothing to do with this PR (you don't seem to use them to make a specific test for this PR's issue? Just improve the readability of existing tests? Or did I miss the one new test? Which only highlights the problem tbh 😅). It's annoying, but I think having small, self-contained PRs is critical later when chasing up some bug or design decision...

Then, ideally, you would use the new infra to test this PR's fixes, but I don't really know how challenging that is. Seems significant. 😅

Comment (brisvag) [reply napari/layers/shapes/_tests/test_shapes_mouse_bindings.py]: I can revert if you prefer and make a separate PR. The main changes are just for readability, though I did change 1 or 2 minor things to improve some tests. All unrelated to this, except the `layer.scale_factor = 0.001` in the layer creation fixture.

File: napari/layers/shapes/_shapes_mouse_bindings.py [hunk 3]
```diff
@@ -205,7 +205,7 @@ def add_rectangle(layer: Shapes, event: MouseEvent) -> None:
     event: MouseEvent
         A proxy read only wrapper around a vispy mouse event.
     """
-    size = layer._vertex_size * layer.scale_factor / 4
+    size = layer._normalized_vertex_radius / 2
```
Comment (Czaki) [inline napari/layers/shapes/_shapes_mouse_bindings.py:221]: here is division by 2 and in definition is division by 2. is it mistake?

Comment (brisvag) [reply napari/layers/shapes/_shapes_mouse_bindings.py:221]: it used to be `/ 4`, so it adds up. Why exactly it was `/ 4` I'm not sure, I haven't digged into that, but it's unchanged and it works :P

File: napari/layers/shapes/_tests/test_shapes_mouse_bindings.py (hunk 2 above)
Comment (jni) [reply napari/layers/shapes/_tests/test_shapes_mouse_bindings.py]: I would in fact prefer... 🙏 I've done a lot of git archeology in my time and big unrelated changes are always annoying to come across.

Comment (brisvag) [reply napari/layers/shapes/_tests/test_shapes_mouse_bindings.py]: Ok, reverted except for the 1 needed change and another where the test was essentially asserting the wrong thing.

Comment (brisvag) [reply napari/layers/shapes/_tests/test_shapes_mouse_bindings.py]: Pulled out the changes to #5917!

**Comment by brisvag** [comment]
Finally all tests passing, will mark as ready!

**Pull request merged**